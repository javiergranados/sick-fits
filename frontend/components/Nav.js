import Link from 'next/link';
import User from './User';
import Signout from './Signout';
import * as S from './styles/Nav';

const Nav = () => (
  <User>
    {data => (
      <S.Nav>
        <Link href="/items">
          <a>Shop</a>
        </Link>
        {data && data.me && (
          <>
            <Link href="/sell">
              <a>Sell</a>
            </Link>
            <Link href="/orders">
              <a>Orders</a>
            </Link>
            <Link href="/me">
              <a>Account</a>
            </Link>
            <Signout />
          </>
        )}
        {(!data || !data.me) && (
          <Link href="/signup">
            <a>Sign in</a>
          </Link>
        )}
      </S.Nav>
    )}
  </User>
);

export default Nav;
