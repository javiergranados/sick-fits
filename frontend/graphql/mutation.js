import gql from 'graphql-tag';

const CREATE_ITEM = gql`
  mutation createItem($title: String!, $description: String!, $price: Int!, $image: String, $largeImage: String) {
    createItem(title: $title, description: $description, price: $price, image: $image, largeImage: $largeImage) {
      id
    }
  }
`;

const UPDATE_ITEM = gql`
  mutation updateItem($id: ID!, $title: String, $description: String, $price: Int) {
    updateItem(id: $id, title: $title, description: $description, price: $price) {
      id
      title
      description
      price
    }
  }
`;

const DELETE_ITEM = gql`
  mutation deleteItem($id: ID!) {
    deleteItem(id: $id) {
      id
      title
    }
  }
`;

const SIGN_UP = gql`
  mutation signup($email: String!, $name: String!, $password: String!) {
    signup(email: $email, name: $name, password: $password) {
      id
      email
      name
    }
  }
`;

const SIGN_IN = gql`
  mutation signin($email: String!, $password: String!) {
    signin(email: $email, password: $password) {
      id
      email
      name
      password
    }
  }
`;

const SIGN_OUT = gql`
  mutation signout {
    signout {
      message
    }
  }
`;

const REQUEST_RESET = gql`
  mutation requestReset($email: String!) {
    requestReset(email: $email) {
      message
    }
  }
`;

export { CREATE_ITEM, UPDATE_ITEM, DELETE_ITEM, SIGN_UP, SIGN_IN, SIGN_OUT, REQUEST_RESET };
