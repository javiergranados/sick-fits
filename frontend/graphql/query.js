import gql from 'graphql-tag';
import { perPage } from '../config';

const GET_ITEMS = gql`
  query($skip: Int = 0, $first: Int = ${perPage}) {
    items(skip: $skip, first: $first, orderBy: createdAt_DESC) {
      id
      title
      price
      description
      image
      largeImage
    }
  }
`;

const GET_ITEM = gql`
  query($id: ID!) {
    item(where: { id: $id }) {
      id
      title
      description
      price
      largeImage
    }
  }
`;

const PAGINATION = gql`
  {
    itemsConnection {
      aggregate {
        count
      }
    }
  }
`;

const CURRENT_USER = gql`
  {
    me {
      id
      name
      email
      permissions
    }
  }
`;

const GET_ALL_USERS = gql`
  {
    users {
      id
      name
      email
      permissions
    }
  }
`;

export { GET_ITEMS, GET_ITEM, PAGINATION, CURRENT_USER, GET_ALL_USERS };
