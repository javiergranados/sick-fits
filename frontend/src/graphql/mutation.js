import gql from 'graphql-tag';

const CREATE_ITEM = gql`
  mutation createItem($title: String!, $description: String!, $price: Int!, $image: String, $largeImage: String) {
    createItem(title: $title, description: $description, price: $price, image: $image, largeImage: $largeImage) {
      id
    }
  }
`;

const UPDATE_ITEM = gql`
  mutation updateItem($id: ID!, $title: String, $description: String, $price: Int) {
    updateItem(id: $id, title: $title, description: $description, price: $price) {
      id
      title
      description
      price
    }
  }
`;

const DELETE_ITEM = gql`
  mutation deleteItem($id: ID!) {
    deleteItem(id: $id) {
      id
    }
  }
`;

export { CREATE_ITEM, UPDATE_ITEM, DELETE_ITEM };
