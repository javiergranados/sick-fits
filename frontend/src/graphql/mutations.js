import gql from 'graphql-tag';

const CREATE_ITEM = gql`
  mutation createItem($title: String!, $description: String!, $price: Int!, $image: String, $largeImage: String) {
    createItem(title: $title, description: $description, price: $price, image: $image, largeImage: $largeImage) {
      id
    }
  }
`;

export { CREATE_ITEM };
