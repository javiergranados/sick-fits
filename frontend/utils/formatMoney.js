export default function(amount) {
  const options = {
    style: 'currency',
    currency: 'EUR',
    minimumFractionDigits: 2,
  };
  // if its a whole, dollar amount, leave off the .00
  if (amount % 100 === 0) options.minimumFractionDigits = 0;
  const formatter = new Intl.NumberFormat('it', options);
  return formatter.format(amount / 100);
}
// number.toLocaleString('it', { style: 'currency', currency: 'EUR' }) : '0,00 â‚¬';
