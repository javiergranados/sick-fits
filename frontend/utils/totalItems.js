export default function(cart) {
  return cart.reduce((prev, next) => prev + next.quantity, 0);
}
