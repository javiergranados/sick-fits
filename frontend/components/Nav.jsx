import { useCallback } from 'react';
import Link from 'next/link';
import { useMutation } from '@apollo/react-hooks';
import User from './User';
import CartCount from './CartCount';
import Signout from './Signout';
import { TOGGLE_CART } from '../graphql/mutation';
import * as S from './styles/Nav';

const Nav = () => {
  const [toggleCart] = useMutation(TOGGLE_CART);
  const handleClick = useCallback(toggleCart, []);

  return (
    <User>
      {({ me }) => (
        <S.Nav data-test="nav">
          <Link href="/items">
            <a>Shop</a>
          </Link>
          {me && (
            <>
              <Link href="/sell">
                <a>Sell</a>
              </Link>
              <Link href="/orders">
                <a>Orders</a>
              </Link>
              <Link href="/me">
                <a>Account</a>
              </Link>
              <Signout />
              <button type="button" onClick={handleClick}>
                My Cart
                <CartCount count={me.cart.reduce((prev, next) => prev + next.quantity, 0)} />
              </button>
            </>
          )}
          {!me && (
            <Link href="/signup">
              <a>Sign in</a>
            </Link>
          )}
        </S.Nav>
      )}
    </User>
  );
};

export default Nav;
