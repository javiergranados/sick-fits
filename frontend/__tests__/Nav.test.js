import { mount } from 'enzyme';
import wait from 'waait';
import toJSON from 'enzyme-to-json';
import { MockedProvider } from '@apollo/react-testing';
import { act } from 'react-dom/test-utils';
import Nav from '../components/Nav';
import { CURRENT_USER } from '../graphql/query';
import { fakeUser, fakeCartItem } from '../utils/testUtils';

const notSignedInMocks = [
  {
    request: { query: CURRENT_USER },
    result: { data: { me: null } },
  },
];

const signedInMocks = [
  {
    request: { query: CURRENT_USER },
    result: { data: { me: fakeUser() } },
  },
];

const signedInMocksWithCartItems = [
  {
    request: { query: CURRENT_USER },
    result: {
      data: {
        me: {
          ...fakeUser(),
          cart: [fakeCartItem(), fakeCartItem(), fakeCartItem()],
        },
      },
    },
  },
];

async function getNav(mock) {
  const wrapper = mount(
    <MockedProvider mocks={mock}>
      <Nav />
    </MockedProvider>
  );
  await act(wait);
  wrapper.update();

  return wrapper.find('ul[data-test="nav"]');
}

describe('<Nav />', () => {
  it('renders a minimal nav when signed out', async () => {
    const nav = await getNav(notSignedInMocks);

    expect(toJSON(nav)).toMatchSnapshot();
  });

  it('renders full nav when signed in', async () => {
    const nav = await getNav(signedInMocks);

    expect(nav.children().length).toBe(6);
    expect(nav.text()).toContain('Sign out');
  });

  it('renders the amount of items in the cart', async () => {
    const nav = await getNav(signedInMocksWithCartItems);

    const count = nav.find('div.count');
    expect(toJSON(count)).toMatchSnapshot();
  });
});
