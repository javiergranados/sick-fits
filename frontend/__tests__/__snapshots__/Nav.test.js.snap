// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Nav/> renders a minimal nav when signed out 1`] = `
<ul
  className="Nav-sc-3uwxy3-0 iPyuTS"
  data-test="nav"
>
  <Link
    href="/items"
  >
    <a
      href="/items"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      Shop
    </a>
  </Link>
  <Link
    href="/signup"
  >
    <a
      href="/signup"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      Sign in
    </a>
  </Link>
</ul>
`;

exports[`<Nav/> renders the amount of items in the cart 1`] = `
<div
  className="CartCount__Dot-sc-1ror63v-1 kUvYyW count"
>
  9
</div>
`;
